data/simpy.gram

```diff
@@ -324,10 +324,22 @@ power[expr_ty]:
     | primary 
 primary[expr_ty]:
     | !'await' a=atom !'.' !'(' !'[' { a }
-    | [AWAIT] atom ('.' NAME | '[' slices ']' | '(' [arguments] ')')*
+    | AWAIT a=primary_raw { _Py_Await(a, EXTRA) }
+    | primary_raw
+primary_raw[expr_ty]:
+    | a=primary_raw '.' b=NAME { _Py_Attribute(a, ((expr_ty) b)->v.Name.id, Load, EXTRA) }
+    | a=primary_raw '[' b=slices ']' { _Py_Subscript(a, b, Load, EXTRA) }
+    | a=primary_raw '(' b=[arguments] ')' {
+        _Py_Call(a,
+                 (b) ? ((expr_ty) b)->v.Call.args : NULL,
+                 (b) ? ((expr_ty) b)->v.Call.keywords : NULL,
+                 EXTRA) }
+    | atom
 
-slices: ','.slice+ [',']
-slice: [expression] ':' [expression] [':' [expression]] | expression
+slices[slice_ty]: a=','.slice+ [','] { (asdl_seq_LEN(a) == 1) ? asdl_seq_GET(a, 0) : _Py_ExtSlice(a, p->arena) }
+slice[slice_ty]:
+    | a=[expression] ':' b=[expression] c=[':' d=[expression] { d }] { _Py_Slice(a, b, c, p->arena) }
+    | a=expression { _Py_Index(a, p->arena) }
```

Comment:
Seems the logic in `ast.c` (in `ast_for_trailer()`) is different. If there are multiple slices but none of them have colons, it uses `Index(Tuple(...))` rather than `ExtSlice(...)`. Using the power of PEG and backtracking we could code this by using an extra alternative in `primary_raw`:
```
    | a=primary_raw '[' b=slice_expressions ']' { ... }
    | a=primary_raw '[' b=slices ']' { ... }
```
so that backtracking will happen if there are any colons. Then we need
```
slice_expressions: ','.expression+ [','] { ... }
```
Alternatively we could write a helper function to implement that logic (which may result in less generated code but more code to maintain, in a sense).

Reply:
Oh yeah, for some reason I didn't check to see what `ast.c` does. Sorry!